include/boost/unordered/detail/foa/core.hpp

```diff
@@ -1054,13 +1051,7 @@ struct table_arrays
     initialize_groups(
       arrays.groups(),groups_size,
       std::integral_constant<
-        bool,
-#if BOOST_WORKAROUND(BOOST_LIBSTDCXX_VERSION,<50000)
-      /* std::is_trivially_constructible not provided */
-      boost::has_trivial_constructor<group_type>::value
-#else
-      std::is_trivially_constructible<group_type>::value
-#endif  
+        bool,std::is_trivially_constructible<group_type>::value
```

Comment:
I'm glad you brought this up because I wasn't sure about this change either.

Inadvertently, we removed gcc<5.0 from our testing matrix with the new testing requirements.

Back when I asked Glen how to handle the C++03 drop, he told me:
> Basically when the time comes, don't use anything that is a compensation/workaround for missing or incomplete C++11 language support

Now that I look back on it, I'm not sure if he menat to include library support in there as well but my operating assumption was: no more workarounds period.

Reply:
I think what we can do is, clean up the code so that we don't have any more of the workarounds and if a user a complains, we'll polyfill whatever we need from TypeTraits to make it all work. Otherwise, I'm perfectly content dropping gcc-4 as its `<type_traits>` header isn't up to par yet.

Reply:
It's up to us whether to support gcc 4.8 and 4.9. I think we can afford to drop them in this day and age. Or if not, we can use `std::has_trivial_constructor` instead of `boost::has_trivial_constructor`.

Reply:
E.g. https://github.com/boostorg/variant2/blob/a5cb0582d5669e86ee913f97c001677987023967/include/boost/variant2/variant.hpp#L510-L535

Reply:
Well, apparently updating those testing requirements turned out to be a _mistake_ as now foa segfaults for gcc-4.8. I'm looking into it now, of course.

Reply:
@cmazakas I think copying Variant2's workaround is an excellent solution, as it's not dependent on Boost.TypeTraits.

Reply:
@pdimov, why:
```cpp
std::is_copy_constructible<T>::value && std::has_trivial_copy_constructor<T>::value
```
instead of just:
```cpp
std::has_trivial_copy_constructor<T>::value
```
?
Same with the other traits.

Reply:
I don't remember, but let me try to figure it out. There must have been a reason.

Reply:
I don't really know. I tried a few corner cases and it seems that the two traits agree. Let's just use `has_trivial_copy_constructor` by itself and see if it creates any trouble.

Reply:
Ah, found it. A deleted copy constructor is trivial. However, GCC 4.8 doesn't consider it one, which is why I didn't see it in my tests. 4.9 and above do (https://godbolt.org/z/evoq7daKq, https://godbolt.org/z/TK5cMMe1c).

(Why is a deleted copy constructor trivial? Nobody knows. It was done to keep the copy constructors and assignment operators of `atomic<T>` trivial, supposedly for compatibility with C, but they aren't actually trivial, which necessitated patching this on the library side via [LWG issue 2424](https://www.open-std.org/jtc1/sc22/wg21/docs/lwg-defects.html#2424). Mysterious ways.)

Reply:
Should then the following:
```cpp
#if BOOST_WORKAROUND(BOOST_LIBSTDCXX_VERSION,<50000)
      /* std::is_trivially_default_constructible not provided */
      std::has_trivial_default_constructor<group_type>::value
#else
      std::is_trivially_default_constructible<group_type>::value
#endif 
```
be rewritten as
```cpp
#if BOOST_WORKAROUND(BOOST_LIBSTDCXX_VERSION,<50000)
      /* std::is_trivially_default_constructible not provided */
      std::is_default_constructible<group_type>::value&&
      std::has_trivial_default_constructor<group_type>::value
#else
      std::is_trivially_default_constructible<group_type>::value
#endif 
```
or do those deletion issues not apply here?

Reply:
GCC 4.9 considers a deleted default constructor trivial (https://godbolt.org/z/Wr6r8qrfx) so yes.

Reply:
Actually, they are required. We seem to erroneously trigger `true` in cases that we shouldn't and we wind up segfaulting in the tests.